File: Test/look_up_table_model.cpp
#include "look_up_table_model.h"

look_up_table_model::look_up_table_model(QObject *parent) : QAbstractTableModel(parent)
{
}

// Create a method to populate the model with data:
void look_up_table_model::populateData(const QList<QString> &contactName,const QList<QString> &contactPhone,const QList<int> &contactnum, const QList<float> &frame_1,const QList<float> &X_Position,const QVector<QList<float>> &Data)
{
    tm_contact_name.clear();
    tm_contact_name = contactName;
    tm_contact_phone.clear();
    tm_contact_phone = contactPhone;
    tm_contact_numb.clear();
    tm_contact_numb = contactnum;
    Frame_1.clear();
    Frame_1 = frame_1;
    tm_X_Position.clear();
    tm_X_Position = X_Position;
    tm_Data.clear();
    tm_Data = Data;
    return;
}

int look_up_table_model::rowCount(const QModelIndex &parent) const
{
    Q_UNUSED(parent);
    return tm_contact_name.length();
}

int look_up_table_model::columnCount(const QModelIndex &parent) const
{
    Q_UNUSED(parent);
    return 25;
}

QVariant look_up_table_model::data(const QModelIndex &index, int role) const
{
    if (!index.isValid() || role != Qt::DisplayRole) {
        return QVariant();
    }
    if (index.column() == 0) {
        return tm_Data[look_up_table_model::X_Position][index.row()];
    } else if (index.column()==1){
        return tm_Data[look_up_table_model::Y_Position][index.row()];
    } else if (index.column()==2){
        return tm_Data[look_up_table_model::Z_Position][index.row()];
    } else if (index.column()==3){
        return tm_Data[look_up_table_model::X_Angel][index.row()];
    } else if (index.column()==4){
        return tm_Data[look_up_table_model::Y_Angel][index.row()];
    } else if (index.column()==5){
        return tm_Data[look_up_table_model::Z_Angel][index.row()];
    } else if (index.column()==6){
        return tm_Data[look_up_table_model::X_Coil][index.row()];
    } else if (index.column()==7){
        return tm_Data[look_up_table_model::Y_Coil][index.row()];
    } else if (index.column()==8){
        return tm_Data[look_up_table_model::Z_Coil][index.row()];
    } else if (index.column()==9){
        return tm_Data[look_up_table_model::Frame_antenne_1][index.row()];
    } else if (index.column()==10){
        return tm_Data[look_up_table_model::Frame_antenne_2][index.row()];
    } else if (index.column()==11){
        return tm_Data[look_up_table_model::Frame_antenne_3][index.row()];
    } else if (index.column()==12){
        return tm_Data[look_up_table_model::Frame_antenne_4][index.row()];
    } else if (index.column()==13){
        return tm_Data[look_up_table_model::Frame_antenne_5][index.row()];
    } else if (index.column()==14){
        return tm_Data[look_up_table_model::Frame_antenne_6][index.row()];
    } else if (index.column()==15){
        return tm_Data[look_up_table_model::Frame_antenne_7][index.row()];
    } else if (index.column()==16){
        return tm_Data[look_up_table_model::Frame_antenne_8][index.row()];
    } else if (index.column()==17){
        return tm_Data[look_up_table_model::Main_antenne_1][index.row()];
    } else if (index.column()==18){
        return tm_Data[look_up_table_model::Main_antenne_2][index.row()];
    }else if (index.column()==19){
        return tm_Data[look_up_table_model::Main_antenne_3][index.row()];
    }else if (index.column()==20){
        return tm_Data[look_up_table_model::Main_antenne_4][index.row()];
    }else if (index.column()==21){
        return tm_Data[look_up_table_model::Main_antenne_5][index.row()];
    }else if (index.column()==22){
        return tm_Data[look_up_table_model::Main_antenne_6][index.row()];
    }else if (index.column()==23){
        return tm_Data[look_up_table_model::Main_antenne_7][index.row()];
    }else if (index.column()==24){
        return tm_Data[look_up_table_model::Main_antenne_8][index.row()];
    }
    return QVariant();
}

QVariant look_up_table_model::headerData(int section, Qt::Orientation orientation, int role) const
{
    if (role == Qt::DisplayRole && orientation == Qt::Horizontal) {
        if (section == 0) {
            return QString("X_Position");
        } else if (section == 1) {
            return QString("Y_Position");
        } else if (section ==2){
            return QString("Z_Position");
        } else if (section == 3){
            return QString("X_Angel");
        } else if (section == 4){
            return QString("Y_Angel");
        } else if (section == 5){
            return QString("Z_Angel");
        } else if (section == 6){
            return QString("X_Coil");
        } else if (section == 7){
            return QString("Y_Coil");
        } else if (section == 8){
            return QString("Z_Coil");
        } else if (section == 9){
            return QString("Frame_1");
        } else if (section == 10){
            return QString("Frame_2");
        } else if (section == 11){
            return QString("Frame_3");
        } else if (section == 12){
            return QString("Frame_4");
        } else if (section == 13){
            return QString("Frame_5");
        } else if (section == 14){
            return QString("Frame_6");
        } else if (section == 15){
            return QString("Frame_7");
        } else if (section == 16){
            return QString("Frame_8");
        } else if (section == 17){
            return QString("Main_1");
        } else if (section == 18){
            return QString("Main_2");
        } else if (section == 19){
            return QString("Main_3");
        } else if (section == 20){
            return QString("Main_4");
        } else if (section == 21){
            return QString("Main_5");
        } else if (section == 22){
            return QString("Main_6");
        } else if (section == 23){
            return QString("Main_7");
        } else if (section == 24){
            return QString("Main_8");
        }
    }
    return QVariant();
}

File: look_up_table_model.py

from PyQt5.QtWidgets import *
from PyQt5.QtCore import *
from PyQt5.QtGui import *

class look_up_table(object):
    def __init__(self,X_position,Y_position,Z_position,X_angle,Y_angle,Z_angle, X_coil, Y_coil, Z_coil,Frame_1,Frame_2,Frame_3,Frame_4,Frame_5,Frame_6,Frame_7,Frame_8,Main_1,Main_2,Main_3,Main_4, Main_5, Main_6, Main_7,Main_8):
        self.X_position = X_position
        self.Y_position = Y_position
        self.Z_position = Z_position
        self.X_angle = X_angle
        self.Y_angle = Y_angle
        self.Z_angle = Z_angle
        self.X_coil = X_coil
        self.Y_coil = Y_coil
        self.Z_coil = Z_coil
        self.Frame_1 = Frame_1
        self.Frame_2 = Frame_2
        self.Frame_3 = Frame_3
        self.Frame_4 = Frame_4
        self.Frame_5 = Frame_5
        self.Frame_6 = Frame_6
        self.Frame_7 = Frame_7
        self.Frame_8 = Frame_8
        self.Main_1 = Main_1
        self.Main_2 = Main_2
        self.Main_3 = Main_3
        self.Main_4 = Main_4
        self.Main_5 = Main_5
        self.Main_6 = Main_6
        self.Main_7 = Main_7
        self.Main_8 = Main_8

class look_up_table_model(QAbstractItemView):

    Row_Batch_Count = 15
    def __init__(self):
        super(look_up_table_model, self).__init__()
        self.headers = ['X-position', 'Y-position', 'Z-position', 'X-angle','Y-angle','Z-angle', 'coil X', 'coil Y', 'coil Z','Frame 1','Frame 2', 'Frame 3', 'Frame 4', 'Frame 5','Frame 6','Frame 7', 'Frame 8', 'Main 1', 'Main 2','Main 3','Main 4', 'Main 5', 'Main 6', 'Main 7','Main 8']
        self.tm_look_up_table = []
        self.rowsLoaded = look_up_table_model.Row_Batch_Count

    def rowCount(self,index=QModelIndex()):
        if not self.tm_look_up_table:
            return 0
        if len(self.tm_look_up_table)<= self.rowsLoaded:
            return len(self.tm_look_up_table)
        else:
            return self.rowsLoaded

    def canFetchMore(self,index=QModelIndex()):
        if len(self.tm_look_up_table)> self.rowsLoaded:
            return True
        else:
            return False

    def fetchMore(self,index=QModelIndex()):
        reminder = len(self.tm_look_up_table)-self.rowsLoaded
        itemsToFetch = min(reminder,look_up_table_model.Row_Batch_Count)
        self.beginInsertRows(QModelIndex(),self.rowsLoaded,self.rowsLoaded+itemsToFetch-1)
        self.rowsLoaded += itemsToFetch
        self.endInsertRows()

    def add_look_up_table(self,look_up_table):
        self.beginResetModel()
        self.tm_look_up_table.append(look_up_table)
        self.endResetModel()

    def columnCount(self,index=QModelIndex()):
        return len(self.headers)

    def data(self,index,role=Qt.DisplayRole):
        col= index.column()
        look_up_table = self.tm_look_up_table[index.row()]
        if role == self.Qt.DisplayRole:
            if col == 0:
                return QVariant(look_up_table.X_position)
            elif col ==1:
                return QVariant(look_up_table.Y_position)
            elif col ==2:
                return QVariant(look_up_table.Z_position)
            elif col == 3:
                return QVariant(look_up_table.X_angel)
            elif col ==4:
                return QVariant(look_up_table.Y_angel)
            elif col ==5:
                return QVariant(look_up_table.Z_angel)
            elif col == 6:
                return QVariant(look_up_table.X_coil)
            elif col ==7:
                return QVariant(look_up_table.Y_coil)
            elif col ==8:
                return QVariant(look_up_table.Z_coil)
            elif col == 9:
                return QVariant(look_up_table.Frame_1)
            elif col ==10:
                return QVariant(look_up_table.Frame_2)
            elif col ==11:
                return QVariant(look_up_table.Frame_3)
            elif col == 12:
                return QVariant(look_up_table.Frame_4)
            elif col ==13:
                return QVariant(look_up_table.Frame_5)
            elif col ==14:
                return QVariant(look_up_table.Frame_6)
            elif col ==15:
                return QVariant(look_up_table.Frame_7)
            elif col ==16:
                return QVariant(look_up_table.Frame_8)
            elif col ==17:
                return QVariant(look_up_table.Main_1)
            elif col ==18:
                return QVariant(look_up_table.Main_2)
            elif col ==19:
                return QVariant(look_up_table.Main_3)
            elif col == 20:
                return QVariant(look_up_table.Main_4)
            elif col ==21:
                return QVariant(look_up_table.Main_5)
            elif col ==22:
                return QVariant(look_up_table.Main_6)
            elif col ==23:
                return QVariant(look_up_table.Main_7)
            elif col ==24:
                return QVariant(look_up_table.Main_8)
            return QVariant()

    def headerData(self,section,orientation,role=Qt.DisplayRole):
        if role!= Qt.DisplayRole:
            return QVariant()
        if orientation== Qt.Horizontal:
            return QVariant(self.headers[section])
        return QVariant(int(section+1))
